### ISSUE TITLE
TokenCache Items empty on device, works on npm run emulate

### ISSUE THREAD

**Issue by robinsonlam** [AzureAD/azure-activedirectory-library-for-cordova#129]
### Issue Description
Hey, I'm debugging my cordova app which I've uploaded to Adobe Phonegap Build. When I run the app on my iPad Air 2, I found that tokenCache.readItems gives me an empty array.

Is anyone else having this problem?

### Repro Environment Details
1. Installed plugin version:  ~0.8.1
2. Target Platform: iOS
3. Test device and OS version: iPad Air 2
4. Are you using Cordova from command line or special Cordova IDE (Visual Studio Cordova Tools, etc), what version? Phonegap cli-6.3.1

**Comment by vladimir-kotikov** [comment]
Hey, @robinsonlam, thanks for reporting! The symptoms sound similar to something that is described in #101 - there might be an issue with signing and application access groups, but I'd like to double check.

Also could you provide versions of iOS on your device and cordova-ios platform in your project. Also, what is the content of `emulate` script in your `package.json`

**Comment by robinsonlam** [comment]
Hey @vladimir-kotikov let me add a bit more info

I can get the cache working if I build the project through `cordova build ios` or` cordova emulate ios `to update the ios project in platform/ios. Then I run the xcodeproject and Run the project on the simulator.

I wonder how this plugin persist the cache on iOS devices?

I'm also building the project through phonegap build.

**Comment by vladimir-kotikov** [comment]
Thanks for clarification! I'm going to try reproducing this w/ latest cordova-ios available on phonegap build and iOS 9/10 but still, it'd be great if you can provide cordova-ios and your device's OS versions.

**Comment by vladimir-kotikov** [comment]
@robinsonlam, actually I might now where is the problem - if you're using phonegap 6.3.1 then it's likely that you have cordova-ios version ~4.1.1 - in this version entitlements, that are necessary for proper functioning of the cache, are set up by plugin's hook `before_install` hook (see https://github.com/AzureAD/azure-activedirectory-library-for-cordova/blob/master/plugin.xml#L87-L88). However, AFAIK phonegap build doesn't support plugin hooks by design, and your app ends up not configured properly (in local development hooks are working - that's why cache is not empty when you building the app locally)

Since 4.3.0 cordova-ios has a way to set up entitlements using `edit-config` tag so it's very likely that upgrading to 4.3.0 or later would resolve your problem. Notice that you'll also need to install the latest version of this plugin (0.9.0 which will be released in a couple of days, or install it directly from this repo) that sets entitlements in a proper way.

**Comment by robinsonlam** [comment]
@vladimir-kotikov Thanks for that! The only problem with this is that when I've uploaded my build to Phonegap build using 4.3.0, the build fails giving me a cert/profile error, not sure if you know anything about this?

**Comment by robinsonlam** [comment]
hey @vladimir-kotikov thanks for the prompt responses,
I've checked it out and it seems to work when I emulate it, I haven't tested on the actual device yet so I can't confirm it's 100% working haha. Is 0.9.0 not released yet on the Cordova plugins?

**Comment by vladimir-kotikov** [comment]
So you have upgraded to cordova-ios 4.3.0? As for the plugin, 0.9.0 has been released just a few hours ago, so you can now install it from NPM.

**Comment by robinsonlam** [comment]
@vladimir-kotikov Thanks for the response over the weekend, could I ask how I would update to cordova-ios 4.3.0?

When I had last updated to PhoneGap CLI 6.5.0, I would get some certificate errors and wouldn't be able to build my app. Is there a way around this?

It's still not working on the device but all I've done is change the `config.xml` to use ~0.9.0 of `cordova-plugin-ms-adal`

**Comment by robinsonlam** [comment]
Hey @vladimir-kotikov Thanks your the response over the weekend, I've updated PhoneGap CLI to 6.5.0 which has given me ios version 4.3.1,

I've done:
`<plugin name="cordova-plugin-ms-adal" spec="~0.9.0" />`

in my `config.xml` but I'm still getting the same results on the device.

Am I installing this correctly?

**Comment by vladimir-kotikov** [comment]
Yeah, that's right. Just to make sure that you have cordova-ios@4.3.1, could you check build.phonegap.com/apps and see, which version of the platform is displayed in app details: 

![screen shot 2017-02-15 at 18 18 51](https://cloud.githubusercontent.com/assets/3857604/22980820/ab4c8c80-f3ab-11e6-98b4-ee3c89e7f165.png)

**Comment by robinsonlam** [comment]
Yep, it's 4.3.1, I'm wondering if there are some permissions that are needed? maybe localStorage or something?

Thanks for your help @vladimir-kotikov I really appreciate it :)

<img width="324" alt="screen shot 2017-02-17 at 10 12 26 am" src="https://cloud.githubusercontent.com/assets/7957019/23045671/c6473e1e-f4f9-11e6-885a-ac759317fc67.png">
<img width="724" alt="screen shot 2017-02-17 at 10 12 34 am" src="https://cloud.githubusercontent.com/assets/7957019/23045673/c679f692-f4f9-11e6-97d3-05b687433b89.png">

**Comment by vladimir-kotikov** [comment]
@daserge, could you also please try reproducing this?

**Comment by daserge** [comment]
@vladimir-kotikov @robinsonlam I confirm the issue - token cache does not work, investigating it.

**Comment by vladimir-kotikov** [comment]
Thanks!

**Comment by robinsonlam** [comment]
Hey @vladimir-kotikov @daserge Thanks guys!

Was just wondering if there was any news on this and if you guys need any more information?

**Comment by daserge** [comment]
@robinsonlam As far as I could see from my testing the issue is that `archived-expanded-entitilements.xcent` does not contain `keychain-access-groups`->`com.microsoft.adalcache` as opposed to CLI builds.
So `plugin.xml`' `config-file` does not seem work for PhoneGap Build (or there's a configuration issue) - the behavior was the same when I tested [phonegap-plugin-push](https://github.com/phonegap/phonegap-plugin-push/blob/master/plugin.xml#L98) - i.e. there is no `aps-environment` in entitlements after build.

@robinsonlam are you using Debug or Distribution certificate for PhoneGap Build signing?

**Comment by robinsonlam** [comment]
@daserge It's signed by the Enterprise Developer Program cert for iOS, 
is there any way to enable `keychain-access-groups`->`com.microsoft.adalcache` through a plugin?

**Comment by robinsonlam** [comment]
Hey @daserge, thanks for the help.

Do you mind if I ask if there's any news on this?

**Comment by daserge** [comment]
@robinsonlam I propose to ask [this question](https://github.com/AzureAD/azure-activedirectory-library-for-cordova/issues/129#issuecomment-281654532) on https://forums.adobe.com/community/phonegap/build

Also there's a related thread https://forums.adobe.com/message/9191173#9191173 with a workaround of re-signing the .ipa manually so you might need to add this to the entitlements:

	<key>keychain-access-groups</key>
	<array>
		<string>TEAM_ID.com.microsoft.adalcache</string>
	</array>

Alternatively you can do a local build with Cordova or PhoneGap CLI.

**Comment by robinsonlam** [comment]
@daserge Thanks, we'll be taking it to manual build, looks like it works there.

Could you tell me how you found that `archived-expanded-entitilements.xcent` did not contain `keychain-access-groups` -> `com.microsoft.adalcache` ?

I'm having trouble with another plugin and feel like it may be the same thing

**Comment by daserge** [comment]
@robinsonlam `archived-expanded-entitilements.xcent` can be found inside the `.ipa` archive - you can rename it to zip and extract.
